### PR TITLE
Use spawnSync instead of execaSync in `git.ts`

### LINKED ISSUE #887: ESM errors when building.
### What version of `starlight` are you using?

0.11.0

### What version of `astro` are you using?

3.3.0

### What package manager are you using?

yarn

### What operating system are you using?

Mac

### What browser are you using?

Chrome

### Describe the Bug

When I try to build with `astro build` I get:
```
dist/_astro/MagnifingImage.astro_astro_type_script_index_0_lang.42c36336.js   0.12 kB │ gzip:  0.13 kB
dist/_astro/Tabs.astro_astro_type_script_index_0_lang.ea48f09d.js             0.99 kB │ gzip:  0.49 kB
dist/_astro/hoisted.dbfb2330.js                                               6.03 kB │ gzip:  2.43 kB
dist/_astro/ui-core.42dfa298.js                                              51.45 kB │ gzip: 16.83 kB
✓ built in 648ms
Completed in 664ms.


 generating static routes 
 error   Named export 'execaSync' not found. The requested module 'execa' is a CommonJS module, which may not support all module.exports as named exports.
  CommonJS modules can always be imported via the default export, for example using:
  
  import pkg from 'execa';
  const { execaSync } = pkg;
  
file:///Volumes/Mattia%20SSD/web-development.nosync/monarchy-projects/astro-air/packages/documentation/dist/chunks/pages/404_2ceafa5d.mjs:8
import { execaSync } from 'execa';
         ^^^^^^^^^
SyntaxError: Named export 'execaSync' not found. The requested module 'execa' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'execa';
const { execaSync } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:190:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
package.json:
```
{
  "name": "documentation",
  "type": "module",
  "version": "0.0.1",
  "scripts": {
    "dev": "astro dev",
    "start": "astro dev",
    "build": "astro build",
    "preview": "astro preview",
    "astro": "astro"
  },
  "dependencies": {
    "@astrojs/starlight": "^0.11.0",
    "astro": "^3.3.0",
    "sharp": "^0.32.6"
  },
  "devDependencies": {
    "typescript": "^5.2.2"
  }
}
```

tsconfig is standard.

I cannot give a repro repo right now.

It builds cleanly with bun. Yarn gives this issue, any clues?

### Link to Minimal Reproducible Example

https://google.com

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by kevinzunigacuellar**
#### Description

- Closes #887
- Replace `execa` with `node:child_process`

### CODE REVIEW

**Comment by HiDeoo** [comment]
This is super cool, I remember starting something similar a while back but never finished it. Glad to see someone tackling this :raised_hands:

A few notes:

- I'm not sure this should spawn a new shell so I think `spawnSync` (which accepts a `encoding: 'utf-8'` option to get a string back in `result.stdout`) might be a better choice and would also not change the current behavior.
- `execa` should be removed.
- Maybe we could take this opportunity to remove the `oldest` code path as we do not ever use it (and maybe rename the function? I remember using  `getFileNewestCommitDate` at the time) and also remove `FileNotTrackedError` and replace it by a basic `Error`?
- I remember starting to work on a `git.test.ts` file for this as we started to own a bit more this piece of code. If you think this could be useful, I'm sharing the WIP here in case it could help (this assume `git` is installed on the machine running the tests but I think this is a fair assumption?)

  ```ts
  import { mkdtempSync, writeFileSync } from 'node:fs';
  import { join } from 'node:path';
  import { tmpdir } from 'node:os';
  import { spawnSync } from 'node:child_process';
  import { describe, expect, test } from 'vitest';
  import { getFileNewestCommitDate } from '../../utils/git';

  describe('getFileNewestCommitDate', () => {
    const { commitAllChanges, getFilePath, writeFile } = makeTestRepo();

    test('returns the newest commit date', () => {
      const file = 'updated.md';
      const lastCommitDate = '2023-06-25';

      writeFile(file, 'content 0');
      commitAllChanges('add updated.md', '2023-06-21');
      writeFile(file, 'content 1');
      commitAllChanges('update updated.md', lastCommitDate);

      expectCommitDateToEqual(getFileNewestCommitDate(getFilePath(file)), lastCommitDate);
    });

    test('returns the initial commit date for a file never updated', () => {
      const file = 'added.md';
      const commitDate = '2022-09-18';

      writeFile(file, 'content');
      commitAllChanges('add added.md', commitDate);

      expectCommitDateToEqual(getFileNewestCommitDate(getFilePath(file)), commitDate);
    });

    test('returns the newest commit date for a file with a name that contains a space', () => {
      const file = 'updated with space.md';
      const lastCommitDate = '2021-01-02';

      writeFile(file, 'content 0');
      commitAllChanges('add updated.md', '2021-01-01');
      writeFile(file, 'content 1');
      commitAllChanges('update updated.md', lastCommitDate);

      expectCommitDateToEqual(getFileNewestCommitDate(getFilePath(file)), lastCommitDate);
    });

    test('returns the newest commit date for a file updated the same day', () => {
      const file = 'updated-same-day.md';
      const lastCommitDate = '2023-06-25T14:22:35Z';

      writeFile(file, 'content 0');
      commitAllChanges('add updated.md', '2023-06-25T12:34:56Z');
      writeFile(file, 'content 1');
      commitAllChanges('update updated.md', lastCommitDate);

      expectCommitDateToEqual(getFileNewestCommitDate(getFilePath(file)), lastCommitDate);
    });

    test('throws when failing to retrieve the git history for a file', () => {
      expect(() =>
        getFileNewestCommitDate(getFilePath('../not-a-starlight-test-repo/test.md'))
      ).toThrow(/^Failed to retrieve the git history for file "[/\\-\w ]+\/test\.md" with exit code/);
    });

    test('throws when trying to get the history of a non-existing or untracked file', () => {
      const expectedError =
        /^Failed to retrieve the git history for file "[/\\-\w ]+\/(?:unknown|untracked)\.md" because the file is not tracked by git\.$/;

      writeFile('untracked.md', 'content');

      expect(() => getFileNewestCommitDate(getFilePath('unknown.md'))).toThrow(expectedError);
      expect(() => getFileNewestCommitDate(getFilePath('untracked.md'))).toThrow(expectedError);
    });
  });

  function expectCommitDateToEqual(commitDate: CommitDate, expectedDateStr: ISODate) {
    const expectedDate = new Date(expectedDateStr);

    expect(commitDate).toStrictEqual({
      date: expectedDate,
      timestamp: expectedDate.getTime() / 1000,
    });
  }

  function makeTestRepo() {
    const repoPath = mkdtempSync(join(tmpdir(), 'starlight-test-git-'));

    function runInRepo(command: string, args: string[], env: NodeJS.ProcessEnv = {}) {
      const result = spawnSync(command, args, { cwd: repoPath, env });

      if (result.status !== 0) {
        throw new Error(`Failed to execute test repository command: '${command} ${args.join(' ')}'`);
      }
    }

    // Configure git specifically for this test repository.
    runInRepo('git', ['init']);
    runInRepo('git', ['config', 'user.name', 'starlight-test']);
    runInRepo('git', ['config', 'user.email', 'starlight-test@example.com']);
    runInRepo('git', ['config', 'commit.gpgsign', 'false']);

    return {
      // The `dateStr` argument should be in the `YYYY-MM-DD` or `YYYY-MM-DDTHH:MM:SSZ` format.
      commitAllChanges(message: string, dateStr: ISODate) {
        const date = dateStr.endsWith('Z') ? dateStr : `${dateStr}T00:00:00Z`;

        runInRepo('git', ['add', '-A']);
        // This sets both the author and committer dates to the provided date.
        runInRepo('git', ['commit', '-m', message, '--date', date], { GIT_COMMITTER_DATE: date });
      },
      getFilePath(name: string) {
        return join(repoPath, name);
      },
      writeFile(name: string, content: string) {
        writeFileSync(join(repoPath, name), content);
      },
    };
  }

  type ISODate =
    | `${number}-${number}-${number}`
    | `${number}-${number}-${number}T${number}:${number}:${number}Z`;

  type CommitDate = ReturnType<typeof getFileNewestCommitDate>;
  ```

**Comment by kevinzunigacuellar** [comment]
Thanks for the review Hideo 💜 . I added most of your suggested changes, and cleaned up the logic a bit. However, I seem to have broken a test 😆, which I am not sure how to fix. It seems that my root file is pointing to the test directory but every time I try changing root path in the `test-config` I break the entire thing 💥. Do you have any ideas?

**Comment by HiDeoo** [comment]
I don't have a computer to actually run the code but from the changes, I would say that reverting the changes to `getLastUpdated()` (minus the name and argument changes to `getFileCommitDate()` ofc) should fix the issue as you removed the try-catch block that we would still need to ignore the error and I think that in this test, the path is invalid as it's based on the actual docs (that's also one of the reason I started working on the dedicated tests I linked earlier so we can easily test the git part in isolation).

**Comment by kevinzunigacuellar** [comment]
Thank you Hideo, it was the error catch 😄 . I was able to add your tests in `git.test.ts` 💜

File: packages/starlight/utils/route-data.ts [hunk 1]
```diff
@@ -70,17 +70,24 @@ function getToC({ entry, locale, headings }: PageProps) {
...
+		let date: Date | undefined;
+		try {
+			date =
+				frontmatterLastUpdated instanceof Date
+					? frontmatterLastUpdated
+					: getNewestCommitDate(currentFilePath);
+		} catch {
+			// If the git command fails, ignore the error and continue.
 		}
 		return date;
```
Comment (delucis) [inline packages/starlight/utils/route-data.ts:87]: If we’re doing it this way, I guess we can do:
```suggestion
		try {
			return frontmatterLastUpdated instanceof Date
					? frontmatterLastUpdated
					: getNewestCommitDate(currentFilePath);
		} catch {
			// If the git command fails, ignore the error.
			return undefined;
		}
```

**Review by delucis — COMMENTED** [review]
Woo this looks great. Thanks @kevinzunigacuellar and thank you @HiDeoo for your very insightful review 💖 

Left one small refactor suggestion, but overall, the code looks good to me.

One thought: have we tested this on Windows? We don’t currently use the last-updated feature on the docs site and building the Starlight docs is the only “test” we currently run on Windows. Would be good to be sure that’s working.

**Review by lilnasy — APPROVED** [review]
Haven't looked at the code but the result looks great to me!
<details>
<summary>steps chris told me to follow</summary>
image
</details>

<details>
<summary>result</summary>
image

</details>

**Comment by hippotastic** [comment]
Also works well for me on Windows 10!

**Comment by at-the-vr** [comment]
Working for me as well on Win 10, I went across the translations found no Issues. Found the locale translation interesting and that's without any hardcoding 🔥🔥
| Locale  | Last Updated Image |
| ------------- | ------------- |
| Deutsch  | image  |
| Japanese & Korean  | image|

There are 13 more languages but these 3 caught my attention 😅. Overall it works 😄

**Review by delucis — APPROVED** [review]
Let’s go! Thanks again @kevinzunigacuellar and thanks everyone for testing this out 💖

**Pull request merged**